// @keepkey/keepkey.js/dist/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var keepkeyManager_1 = require("./keepkeyManager");
exports.KeepKeyManager = keepkeyManager_1.default;
var keepkey_1 = require("./keepkey");
exports.KeepKey = keepkey_1.default;
// Export our device interfaces
var webUSBDevice_1 = require("./webUSBDevice");
exports.WebUSBDevice = webUSBDevice_1.default;
// export { default as hidDevice } from './hidDevice // When it exists
// Export all proto message classes
var messages_pb_1 = require("./kkProto/messages_pb");
exports.Messages = messages_pb_1.default;
var types_pb_1 = require("./kkProto/types_pb");
exports.Types = types_pb_1.default;
var exchange_pb_1 = require("./kkProto/exchange_pb");
exports.Exchange = exchange_pb_1.default;
// Export constant functions and utils
__export(require("./ethereum"));
__export(require("./utils"));
// Export a map of number MessageType enums to their respective keepkey proto class
var messageTypeRegistry_1 = require("./messageTypeRegistry");
exports.messageTypeRegistry = messageTypeRegistry_1.default;
